File: mapsindoors_android/android/src/main/kotlin/com/mapspeople/mapsindoors/models/Filter.kt
package com.mapspeople.mapsindoors.models

import com.google.gson.annotations.SerializedName
import com.mapsindoors.core.MPFilter
import com.mapsindoors.core.MPMapExtend
import com.mapsindoors.core.models.MPLatLngBounds

/**
 * Used to deserialize MPFilter from dart, and converting it into MPFilter in the MapsIndoors SDK
 */
data class Filter(
@SerializedName("take") private val mTake : Int?,
@SerializedName("skip") private val mSkip : Int?,
@SerializedName("categories") private val mCategories : List<String>?,
@SerializedName("locations") private val mLocations : List<String>?,
@SerializedName("types") private val mTypes : List<String>?,
@SerializedName("parents") private val mParents : List<String>?,
@SerializedName("mapExtend") private val mMapExtend : Bounds?,
@SerializedName("geometry") private val mGeometry : Bounds?,
@SerializedName("floorIndex") private val mFloorIndex : Int?,
@SerializedName("depth") private val mDepth : Int?,
@SerializedName("ignoreLocationSearchableStatus") private val mIgnoreLocationSearchableStatus : Boolean?,
@SerializedName("ignoreLocationActiveStatus") private val mIgnoreLocationActiveStatus : Boolean?,
) {
    fun toMPFilter() : MPFilter {
        val builder = MPFilter.Builder()
        mTake?.let(builder::setTake)
        mSkip?.let(builder::setSkip)
        mCategories?.let(builder::setCategories)
        mLocations?.let(builder::setLocations)
        mTypes?.let(builder::setTypes)
        mParents?.let(builder::setParents)
        mMapExtend?.let {
            builder.setMapExtend(MPMapExtend(it.southWest.latLng, it.northEast.latLng))
        }
        mGeometry?.let {
            val mpBounds = MPLatLngBounds(mGeometry.southWest.latLng, mGeometry.northEast.latLng)
            builder.setGeometry(MPFilter.Geometry(mpBounds.center.lat, mpBounds.center.lng, MPMapExtend(mpBounds)))
        }
        mFloorIndex?.let(builder::setFloorIndex)
        mDepth?.let(builder::setDepth)
        mIgnoreLocationActiveStatus?.let(builder::setIgnoreLocationActiveStatus)
        mIgnoreLocationSearchableStatus?.let(builder::setIgnoreLocationSearchableStatus)
        return builder.build()

    }
}

File: mapsindoors_ios/ios/Classes/Models/Filter.swift
//
//  Filter.swift
//  mapsindoors_ios
//
//  Created by Martin Hansen on 20/02/2023.
//

import Foundation
import MapsIndoors
import MapsIndoorsCore

public class Filter: Codable {
    public func toMPFilter( jsonString: String ) -> (MPFilter)
    {
        let queryObj: Filter = try! JSONDecoder().decode(Filter.self, from: Data(jsonString.utf8))

        let result = MPFilter()
            
        if(queryObj.take != nil)
        {
            result.take = queryObj.take!
        }
        if(queryObj.skip != nil)
        {
            result.skip = queryObj.skip!
        }
        if(queryObj.floorIndex != nil)
        {
            result.floorIndex = NSNumber(integerLiteral: queryObj.floorIndex!)
        }
        if(queryObj.categories != nil)
        {
            result.categories = queryObj.categories!
        }
        if(queryObj.locations != nil)
        {
            result.locations = queryObj.locations!
        }
        if(queryObj.types != nil)
        {
            result.types = queryObj.types!
        }
        if(queryObj.parents != nil)
        {
            result.parents = queryObj.parents!
        }
        if(queryObj.mapExtend != nil)
        {
            let sw = CLLocationCoordinate2D.init(latitude: CLLocationDegrees(floatLiteral: queryObj.mapExtend!.southwest.latitude),
                                                 longitude: CLLocationDegrees(floatLiteral: queryObj.mapExtend!.southwest.longitude))
            let ne = CLLocationCoordinate2D.init(latitude: CLLocationDegrees(floatLiteral: queryObj.mapExtend!.northeast.latitude),
                                                 longitude: CLLocationDegrees(floatLiteral: queryObj.mapExtend!.northeast.longitude))
            result.bounds = MPGeoBounds(southWest: sw, northEast: ne)
        }
//        if(queryObj.geometry != nil)
//        {
//            result.geometry = queryObj.geometry!
//        }
        if(queryObj.floorIndex != nil)
        {
            result.floorIndex = NSNumber(integerLiteral: queryObj.floorIndex!)
        }
        if(queryObj.depth != nil)
        {
            result.depth = queryObj.depth!
        }
        if(queryObj.ignoreLocationSearchableStatus != nil)
        {
            result.ignoreSearchableStatus = queryObj.ignoreLocationSearchableStatus!
        }
        if(queryObj.ignoreLocationActiveStatus != nil)
        {
            result.ignoreActiveStatus = queryObj.ignoreLocationActiveStatus!
        }
        return result
    }
    
    public var take: Int?
    public var skip: Int?
    public var categories: [String]?
    public var locations: [String]?
    public var types: [String]?
    public var parents: [String]?

    public var mapExtend: geoBounds?
    public var geometry: geoBounds?
    public var floorIndex: Int?
    public var depth: Int?
    public var ignoreLocationSearchableStatus: Bool?
    public var ignoreLocationActiveStatus: Bool?
    
    public struct geoBounds : Codable {
        var northeast: MPPoint
        var southwest: MPPoint
    }
}
